test/fakeintake/server/serverstore/store.go

```diff
@@ -8,131 +8,60 @@
 package serverstore
 
 import (
+	"errors"
 	"fmt"
 	"log"
-	"sync"
 	"time"
 
 	"github.com/prometheus/client_golang/prometheus"
 
 	"github.com/DataDog/datadog-agent/test/fakeintake/api"
 )
 
-// Store implements a thread-safe storage for raw and json dumped payloads
-type Store struct {
-	mutex sync.RWMutex
-
-	rawPayloads map[string][]api.Payload
-
-	NbPayloads *prometheus.GaugeVec
-}
-
-// NewStore initialise a new payloads store
-func NewStore() *Store {
-	return &Store{
-		mutex:       sync.RWMutex{},
-		rawPayloads: map[string][]api.Payload{},
-		NbPayloads: prometheus.NewGaugeVec(prometheus.GaugeOpts{
-			Name: "payloads",
-			Help: "Number of payloads collected by route",
-		}, []string{"route"}),
-	}
-}
-
-// AppendPayload adds a payload to the store and tries parsing and adding a dumped json to the parsed store
-func (s *Store) AppendPayload(route string, data []byte, encoding string, collectTime time.Time) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	rawPayload := api.Payload{
-		Timestamp: collectTime,
-		Data:      data,
-		Encoding:  encoding,
-	}
-	s.rawPayloads[route] = append(s.rawPayloads[route], rawPayload)
-	s.NbPayloads.WithLabelValues(route).Set(float64(len(s.rawPayloads[route])))
-}
-
-// CleanUpPayloadsOlderThan removes payloads older than time
-func (s *Store) CleanUpPayloadsOlderThan(time time.Time) {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	log.Printf("Cleaning up payloads")
-	// clean up raw payloads
-	for route, payloads := range s.rawPayloads {
-		lastInvalidPayloadIndex := -1
-		for i, payload := range payloads {
-			if payload.Timestamp.Before(time) {
-				lastInvalidPayloadIndex = i
-			}
-		}
-		s.rawPayloads[route] = s.rawPayloads[route][lastInvalidPayloadIndex+1:]
-		s.NbPayloads.WithLabelValues(route).Set(float64(len(s.rawPayloads[route])))
-	}
+// Store is the interface for a store that can store payloads and try parsing them
+type Store interface {
+	// AppendPayload adds a payload to the store and tries parsing and adding a dumped json to the parsed store
+	AppendPayload(route string, data []byte, encoding string, collectTime time.Time) error
+	// CleanUpPayloadsOlderThan removes payloads older than the given time
+	CleanUpPayloadsOlderThan(time.Time)
+	// GetRawPayloads returns all raw payloads for a given route
+	GetRawPayloads(route string) []api.Payload
+	// GetRouteStats returns the number of payloads for each route
+	GetRouteStats() map[string]int
+	// Flush flushes the store
+	Flush()
+	// GetMetrics returns the prometheus metrics for the store
+	GetMetrics() []prometheus.Collector
+	// Close closes the store
+	Close()
 }
 
-// GetRawPayloads returns payloads collected for route `route`
-func (s *Store) GetRawPayloads(route string) (payloads []api.Payload) {
-	s.mutex.RLock()
-	defer s.mutex.RUnlock()
-	payloads = make([]api.Payload, len(s.rawPayloads[route]))
-	copy(payloads, s.rawPayloads[route])
-	return payloads
+// NewStore returns a new store
+func NewStore() Store {
```

Comment:
Do we really need this method here? Since it return an `inMemoryStore` I'd except it to be defined next to the inMemoryStore and to return `*inMemoryStore` instead of an interface

Reply:
The current design is to keep the actual implementations as private details. An external package can only access the public interface. Also the NewStore function is the place where we will initialize potential other implementations.